tokio/examples/tinydb.rs

```diff
@@ -0,0 +1,231 @@
+//!     GET foo
+//!     foo = bar
+//!     GET FOOBAR
+//!     error: no key FOOBAR
+//!     SET FOOBAR my awesome string
+//!     set FOOBAR = `my awesome string`, previous: None
+//!     SET foo tokio
+//!     set foo = `tokio`, previous: Some("bar")
+//!     GET foo
+//!     foo = tokio
+//!
+//! Namely you can issue two forms of commands:
+//!
+//! * `GET $key` - this will fetch the value of `$key` from the database and
+//!   return it. The server's database is initially populated with the key `foo`
+//!   set to the value `bar`
+//! * `SET $key $value` - this will set the value of `$key` to `$value`,
+//!   returning the previous value, if any.
+
+#![feature(async_await)]
+#![feature(async_closure)]
+#![deny(warnings, rust_2018_idioms)]
+
+use std::collections::HashMap;
+use std::env;
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+
+use tokio;
+use tokio::codec::Framed;
+use tokio::codec::LinesCodec;
+use tokio::net::TcpListener;
+
+use futures::SinkExt;
+use futures::StreamExt;
+
+/// The in-memory database shared amongst all clients.
+///
+/// This database will be shared via `Arc`, so to mutate the internal map we're
+/// going to use a `Mutex` for interior mutability.
+struct Database {
+    map: Mutex<HashMap<String, String>>,
+}
+
+/// Possible requests our clients can send us
+enum Request {
+    Get { key: String },
+    Set { key: String, value: String },
+}
+
+/// Responses to the `Request` commands above
+enum Response {
+    Value {
+        key: String,
+        value: String,
+    },
+    Set {
+        key: String,
+        value: String,
+        previous: Option<String>,
+    },
+    Error {
+        msg: String,
+    },
+}
+
+#[tokio::main]
+async fn main() {
+    // Parse the address we're going to run this server on
+    // and set up our TCP listener to accept connections.
+    let addr = env::args().nth(1).unwrap_or("127.0.0.1:8080".to_string());
+    let addr = addr.parse::<SocketAddr>().expect("Failed to parse address");
+    let mut listener = TcpListener::bind(&addr).expect("Failed to bind");
+    println!("Listening on: {}", addr);
+
+    // Create the shared state of this server that will be shared amongst all
+    // clients. We populate the initial database and then create the `Database`
+    // structure. Note the usage of `Arc` here which will be used to ensure that
+    // each independently spawned client will have a reference to the in-memory
+    // database.
+    let mut initial_db = HashMap::new();
+    initial_db.insert("foo".to_string(), "bar".to_string());
+    let db = Arc::new(Database {
+        map: Mutex::new(initial_db),
+    });
+
+    loop {
+        match listener.accept().await {
+            Ok((socket, _)) => {
+                // Here's where the meat of the processing in this server happens. First
+                // we see a clone of the database being created, which is creating a
+                // new reference for this connected client to use. Also note the `move`
```

Comment:
this should be updated move isnt needed.

Reply:
Without `move` there the compiler errors with "`db` doesn't live long enough".

Reply:
Sorry what I mean this comment I believe is referring to `listener.incoming().for_each(move || {..})` which was the old way to do this but now that we have async/await db is captured by ref here.